New-HTMLStatusItem customization

jimmy58663 opened the pull request:
Modifications leave all original functionality and syntax in working order. 

- Added functionality to customize the background color of the status items using color hex codes. ('#5f6982')
- Added functionality to customize the font color of the service status and name using color hex codes. ('#5f6982')
- Added functionality to customize icons to any HTML icon that has a hex code consisting of 4 or 5 digits. ('&#x2620')
- Removed CSS classes from status.css that are no longer needed.
- Added CSS class 'icon' to handle formatting of the icons as the old classes did.

PrzemyslawKlys commented on Public/New-HTMLStatusItem.ps1:
```diff
@@ -1,43 +1,72 @@
+        [string]$Percentage = '100%',
+
+        [ValidatePattern('^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$')]
+        [Parameter(ParameterSetName = 'Hex')]
```
For colors you should use what other commands are using. Register-ArgumentCompleter. It allows for pick from 800 colors but also takes '#COLOR' if typed in.

PrzemyslawKlys commented on Public/New-HTMLStatusItem.ps1:
```diff
@@ -1,43 +1,72 @@
-    } elseif ($Icon -eq 'Good') {
-        #$IconType = 'performance'
+    if ($IconHex) {
+        $IconType = $IconHex
```
For icons you should allow a choice from Font Awesome like other commands are doing.

PrzemyslawKlys commented:
I believe we should leave the old option as one ParameterSet, and create fully customizable one with another set. Not sure if it's worth going with limiting it the way it was limited.

jimmy58663 commented:
I was able to update the colors to utilize the $Script:RGBColors and add the FontAwesome functionality for icons. Unfortunately, I am unable to reach FontAwesome from work where I am generating and using these reports. Are you alright with me keeping the $IconHex as a different parameter set? With the updates there are currently 4 parameter sets:

- 'Statusimo' (Original parameters)
- 'FontAwesomeBrands'
- 'FontAwesomeRegular'
- 'FontAwesomeSolid'

If you are looking to deprecate the original parameters once this is implemented, we could write a warning inside the 'Satusimo' parameter set. I will upload changes when I get home tonight.

PrzemyslawKlys commented:
Yes, we're looking to depreciate old original parameters as long as we update Statusimo to new ones. Yes you can use iconHex - how does that work anyway?

jimmy58663 commented:
Changes have all been uploaded in the latest commit.
- Font  Awesome icon support
- $Script:RGBColor support to align with other functions in the module
- Added/changed parameter set names
- Added a deprecation warning for the 'Statusimo' parameter set

The $IconHex works with UTF-8 Unicode character hex codes. All the main browsers support translation of these codes by default and a reference sheet of codes can be found at https://www.w3schools.com/charsets/ref_html_utf8.asp.

PrzemyslawKlys commented:
That is a cool feature. Probably others would have use of it instead of FontAwesome - with the same problem as yours. It might be worth adding it to other places where font awesome is used.

PrzemyslawKlys commented:
I guess we should also convert Font Awesome to use Register-ArgumentCompleter the same way colors do. But this needs fixing all over the place.

jimmy58663 commented:
It is something that may be beneficial to others. From what I read, Font Awesome does have the capability to be used locally if you wanted to do that instead. I believe you would download the latest CSS from them at build time, pushing to the PSGallery, and then in the head of your HTML instead of linking their site you link <the path to your module>\resources\css\css.all. That reference would probably have to be built dynamically on install of the module or something since there are a few locations it could be. I am not sure how it would react to a source that does not exist if you just added all the locations.

The ArgumentCompleter seemed to work fine for me on the Font Awesome parameters. If you were to look into changing all of those autocomplete to standardize I would recommend dynamic parameters instead. I think the autocomplete functionality on them works better than the Register-ArgumentCompleter. If you would like, I could send you an example of how something like that would be implemented.

PrzemyslawKlys commented:
I do know how to build DynamicParams as I use it in PSWinReportingV2. But as far as I know, it's recommended to go for Register-ArgumentCompleter as it's an easier and cleaner way to work. It also allows to build one "source" and register it in multiple functions. 

As for CSS, we are already doing that for other code/libraries.

- https://github.com/EvotecIT/PSWriteHTML/tree/master/Resources/CSS

But we're not doing linking to CSS local files but instead, we push it inline. 

```powershell
New-Html -UseJavaScrriptLinks -UseCssLinks {

}
```

If you don't use code like above all CSS/JS code is added inline. It's controlled by:

- https://github.com/EvotecIT/PSWriteHTML/blob/master/Private/Parameters.Configuration.ps1

It's as simple as definining CssLink/Css or/and JSlink/Js to have offline support. I was just not aware you can download https://use.fontawesome.com/releases/v5.11.2/css/all.css and use it offline. If you can than that is trivial fix...

https://github.com/EvotecIT/PSWriteHTML/blob/cd1b01deee28e3a5481e70d153e01282d02b0cd4/Private/Parameters.Configuration.ps1#L228-L236

Just create all.css in resources, add Css entry in that file and you've got offline support.
